feat: first version of our implementation

ydayagi opened the pull request.

ydayagi committed:
filter paging users

first version of our implementation
it shows:
- plugin independent of DB
- text filtering
- paging
- users and groups validation

Signed-off-by: Yaron Dayagi <ydayagi@redhat.com>

mareklibra commented on plugins/notifications-backend/src/service/router.ts:
```diff
@@ -1,61 +1,160 @@
   });
 
-  router.get('/notifications/count', async (_, response) => {
+  router.get('/notifications/count/:text?', async ({params: {text}}, response) => {
```
For extensibility, better to use:
```suggestion
  router.get('/notifications/count', async (request, response) => {
  const { text } = request.query;
  // ...
```

ydayagi replied:
exactly what i was looking at earlier - how to read the query params.
I must say I don't like it.
Why? express does not allow me to specify which query params i expect.
also I think the query becomes too big.
I would add a body to the GET with an object: ListNotificationsRequest

mareklibra replied:
I am old-school here and agree that we should not use the `body` in a GET handler - https://stackoverflow.com/questions/978061/http-get-with-request-body

mareklibra commented on plugins/notifications-backend/src/service/router.ts:
```diff
@@ -1,61 +1,160 @@
+    
+    var query = dbClient('messages').count('* as CNT');
+    
+    if ( text && text.length > 0 ) {
```
```suggestion
    if ( text ) {
```

Since empty strings are falsy in JS.

ydayagi replied:
i don't understand your point. an empty string has length equal to 0.

mareklibra replied:
both `(text)` and `(text&&text.length>0)` are equal, they are trueish for a non-empty string.

mareklibra commented on plugins/notifications-backend/src/service/router.ts:
```diff
@@ -1,61 +1,160 @@
         msgcount = (count[0]?.CNT as unknown as number) || -1;
       });
+
     response.json({ status: 'ok', count: msgcount });
```
Do we need the `status`? HTTP 200 should be good enough.

ydayagi replied:
we don't need it. it is leftovers.

mareklibra commented on plugins/notifications-backend/src/service/router.ts:
```diff
@@ -1,61 +1,160 @@
   router.post('/notifications', async (request, response) => {
     let msgid = -1;
+
+    var createRequest = request.body as CreateRequest;
```
This forcing of type is dangerous, we have no guarantee of the input content.
Please note, TypeScript checks are static at build-time only.

ydayagi replied:
i saw in your PR that you do that as well. you send the string body as a parameter of type CreateNotificationRequest. I was under the impression that typescript does Unmarshal/deserialize for us.
 should i perform explicit deserialize with a json library?

mareklibra replied:
Nope, typescript does not do that. At runtime, **all of typescript** is gone (removed at the build-time without adding any semantic change to the code).

We can use such re-type only if we are 100% sure, that the left-side is really of that type.
One of mlutiple ways how to do it is in easing the type (i.e. the `CreateRequest` can have its fields optional and not mandatory).

The de-serialization is done by `JSON.parse()` (most probably) but internally by the Express.

Unless we do thorough structure checking here, we should handle the uncertainty later in the code.

mareklibra commented on plugins/notifications-backend/src/service/router.ts:
```diff
@@ -1,61 +1,160 @@
+    var createRequest = request.body as CreateRequest;
+    
+    if (createRequest.targetGroups) {
+      createRequest.targetGroups.forEach(async group => {
```
Can be `string.forEach` ==> error

mareklibra replied:
Oh, sorry. I meant that the `forEach` can be called on something else than an array as expected. So it would be an error. But this line is outdated already.

mareklibra commented on plugins/notifications-backend/src/service/router.ts:
```diff
@@ -1,61 +1,160 @@
+
+    var createRequest = request.body as CreateRequest;
+    
+    if (createRequest.targetGroups) {
```
```suggestion
    if (Array.isArray(createRequest.targetGroups)) {
```

mareklibra commented on plugins/notifications-backend/src/service/router.ts:
```diff
@@ -1,61 +1,160 @@
+      createRequest.targetGroups.forEach(async group => {
+        const groupRef = await catalogClient.getEntityByRef("group:"+group);
+        if ( !groupRef ) {
+          response.json({ status: 'fail', error: 'group does not exist: '+group });
```
I think we should return HTTP 400 instead. We do not need the `status`

mareklibra commented on plugins/notifications-backend/src/service/router.ts:
```diff
@@ -1,61 +1,160 @@
+    }
+
+    if (createRequest.targetUsers) {
+      createRequest.targetUsers.forEach(async user => {
```
Similar comment regarding the `tagetUsers` type.

Please do not pass `async` function to the `forEach`. It's result is a `Promise` which the `Array.forEach` does not wait for.
If we can not query all-at-once, we have to create an array of promises and then use `const resultArray = await Promise.all(myArrayOfPromises)` to wait.

mareklibra commented on plugins/notifications-backend/src/service/router.ts:
```diff
@@ -1,61 +1,160 @@
       .returning('id')
       .then(id => {
         // We should harden the type
         msgid = id as unknown as number;
```
This retyping is dangerous. We can use sort of
```
.insert(dbObj)
.returning<string, { id: string }[]>('id')
```

mareklibra commented on plugins/notifications-backend/src/service/router.ts:
```diff
@@ -1,61 +1,160 @@
   });
-  router.get('/notifications', async (_, response) => {
+  
+  router.get('/notifications/:text?/:pageSize?/:pageNumber?', async ({params: {text, pageSize, pageNumber}}, response) => {
```
please use `request.query` instead

mareklibra commented on plugins/notifications-backend/src/service/router.ts:
```diff
@@ -1,61 +1,160 @@
+    var query = dbClient('messages').select('*');
+
+    if ( text && text.length > 0 ) {
+      query.whereILike('subject', `%${text}%`).orWhereILike('body', `%${text}%`)
```
We should share the logic building query (WHERE clause) with the `/count`.
There is just a few differences (pagination).

ydayagi replied:
sure

mareklibra commented on plugins/notifications-backend/src/service/router.ts:
```diff
@@ -1,61 +1,160 @@
+    var offset = 0
+
+    if ( pageSize && pageNumber ) {
+      pageSizeNum = Number.parseInt(pageSize)
```
Please mind that this can result in the `NaN` value.

ydayagi replied:
of course

ydayagi replied:
but then it will throw and return error, no?

ydayagi replied:
according to docs NaN returns false like undefined

mareklibra commented on plugins/notifications-backend/src/service/router.ts:
```diff
@@ -1,61 +1,160 @@
-      .select('*')
-      .then(body => {
+    
+    var pageSizeNum = 0
```
I think the reasonable default is
- first page (0)
- page size a >0 constant (i.e. 10 or similar)

To list all messages, we can agree on a reasonable combination, i.e. `pageSize == -1`.

ydayagi replied:
i like keeping it simple for the user. but also let the user say what he wants explicitly. defaults require documentation and at the end there's no reasonable default here. it depends on each one's preference.
I want the params to be there only when using paging. otherwise they have to appear

mareklibra commented on plugins/notifications-backend/src/service/router.ts:
```diff
@@ -1,61 +1,160 @@
+    
+    var query = dbClient('messages').select('*');
+
+    if ( text && text.length > 0 ) {
```
```suggestion
    if ( text) {
```

ydayagi replied:
you need to explain this to me

mareklibra commented on plugins/notifications-backend/src/service/router.ts:
```diff
@@ -1,61 +1,160 @@
+    }
+    
+    await query.then(body => {
         messages = body as unknown as string;
```
I am for naming the requested DB columns in the SELECT, typing them and composing the result explicitly instead of blind passing the result.

ydayagi replied:
of course. we have to export a type for the returned body

mareklibra reviewed (changes requested):
Please run `prettier` and `linter`.

I addition to my comments above, we should separate concerns, mixing the business logic with composition of the router makes the function huge and hard to maintain. Writing unit-tests will be much more complicated.

mareklibra commented on plugins/notifications-backend/src/service/router.ts:
```diff
@@ -1,61 +1,160 @@
+
+  if (! await dbClient.schema.hasTable('messages')) {
+    await dbClient.schema.createTable('messages', function (table) {
+      table.uuid("id", {primaryKey: true}).defaultTo(dbClient.fn.uuid());
```
The UUID is a `string`. I saw it re-typed to `number`

ydayagi replied:
retyped where?

mareklibra replied:
I think it's outdated since the time of writing the comment

ydayagi commented:
> Please run `prettier` and `linter`.
> 
> I addition to my comments above, we should separate concerns, mixing the business logic with composition of the router makes the function huge and hard to maintain. Writing unit-tests will be much more complicated.

it is just a draft. i am working on splitting the code into files.
I don't understand what is business logic and composition. i need small words :-)

mareklibra commented on plugins/notifications-backend/src/service/handlers.ts:
```diff
@@ -0,0 +1,111 @@
+
+    let query = dbClient('messages').select('*');
+
+    addFilter(filter, query);
```
nit: Feel free to ignore
If we flip the params to `addFilter(query, filter)`, it will be easier to extend the list in the future.

ydayagi replied:
fixed

mareklibra commented on plugins/notifications-backend/src/service/router.ts:
```diff
@@ -1,62 +1,66 @@
+    const pageNumberNum = Number.parseInt(pageNumber);
+
+    if ( Number.isNaN(pageSizeNum) || Number.isNaN(pageNumberNum) ) {
+      throw new Error('either pageSize or pageNumber is not a number');
```
I think this check is good-enough (I feel like we do not need the `typeof` check above)

ydayagi replied:
I need 2 checks. why? the parameters can be undefined/string/string[]. HTTP allows multiple query parameters with same name. So my first check verifies it is not undefined nor array of strings.
second check is after parsing into int.
compliation fails if I pass pageSize to parseInt without the first check

ydayagi commented on plugins/notifications-backend/src/service/handlers.ts:
```diff
@@ -0,0 +1,111 @@
+export async function createNotification ( dbClient: Knex<any, any>, catalogClient: CatalogClient
+    , req: CreateNotificationRequest
+    ): Promise<{msgid: string}> {
+    if ( req.targetGroups ) {
```
the type of targetGroups is 'string[] | undefined'. therefore, how can it be something else?

ydayagi replied:
if I check the type then better compare to string[]

ydayagi replied:
fixed

ydayagi commented on plugins/notifications-backend/src/service/handlers.ts:
```diff
@@ -0,0 +1,111 @@
+
+    await dbClient('messages')
+      .insert(row)
+      .returning('id')
```
fixed

ydayagi commented on plugins/notifications-backend/src/service/handlers.ts:
```diff
@@ -0,0 +1,111 @@
+        });
+    }
+
+    if ( req.targetUsers ) {
```
fixed

ydayagi commented on plugins/notifications-backend/src/service/handlers.ts:
```diff
@@ -0,0 +1,111 @@
+    , req: CreateNotificationRequest
+    ): Promise<{msgid: string}> {
+    if ( req.targetGroups ) {
+        req.targetGroups.forEach(async group => {
```
fixed

ydayagi commented on plugins/notifications-backend/src/service/db.ts:
```diff
@@ -0,0 +1,40 @@
+
+  // create tables
+  if (! await dbClient.schema.hasTable('messages')) {
+    await dbClient.schema.createTable('messages', function (table) {
```
fixed

mareklibra replied:
It newly seems this is not a nitpick but requirement by the linter.

ydayagi committed:
fix flpath-659
catch DB errors in router handlers
https://issues.redhat.com/browse/FLPATH-659

Signed-off-by: Yaron Dayagi <ydayagi@redhat.com>

mareklibra commented on plugins/notifications-backend/src/service/router.ts:
```diff
@@ -1,62 +1,76 @@
+      );
+    }
+
+    const pageSizeNum = Number.parseInt(pageSize);
```
To make the linter happy:

```suggestion
    const pageSizeNum = Number.parseInt(pageSize, 10);
```

mareklibra commented on plugins/notifications-backend/src/service/router.test.ts:
```diff
@@ -37,12 +42,12 @@ describe('createRouter', () => {
   });
 
-  describe('GET /health', () => {
+  /* describe('GET /health', () => {
```
if you use `xdescribe(...)`, you do not need to comment-out the code. And so the linter will stay happy, i.e. with the line 22

mareklibra replied:
the xdescribe or xtest is a way to disable particular suite/test
